Rails の i18n を理解する

takao0206 opened the pull request:
サンプルアプリの多言語化（英語、日本語対応）を行いました。

takao0206 committed:
機能の追加: 画面上のリンクで英語と日本語を切り替える機能。

takao0206 committed:
機能の追加: 英語と日本語の翻訳機能を追加。

takao0206 committed:
controllerにも翻訳機能を追加。ymlファイルで翻訳の重複があったので削除。

maedana commented on Gemfile:
```diff
@@ -50,12 +50,16 @@ gem 'bootsnap', require: false
 # Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
 # gem "image_processing", "~> 1.2"
 
+gem 'rails-i18n'
```
どういう用途のgemか理解出来てますか? これを入れてあるなら不要なのではと思う記述が入ってるので理解度が気になります。

takao0206 replied:
アドバイス、ありがとうございました。
`rails-i18n`についての理解が不足しておりました。

gemの機能を理解するため、`config/locals/ja.yml`の基本的な日本語翻訳の部分（例えば、曜日）を削除した後、
`rails c`で日本語翻訳されていることを確認しました（下記）。
```
irb(main):006:0> I18n.locale = :ja
irb(main):007:0> I18n.t('date.day_names')
=> ["日曜日", "月曜日", "火曜日", "水曜日", "木曜日", "金曜日", "土曜日"]
```
よって、自分が個別に追加した日本語訳以外は、`ja.yml`から削除したいと思います。

maedana commented on Gemfile:
```diff
@@ -50,12 +50,16 @@ gem 'bootsnap', require: false
 end
 
 group :development do
+  gem 'i18n_generators'
```
ちゃんと`develpment`groupにできてますね :+1:

takao0206 replied:
ありがとうござます。
今後も、gemの目的や使い方を理解して、group分けしたいと思います。

maedana commented on app/controllers/books_controller.rb:
```diff
@@ -25,7 +25,7 @@ def create
     respond_to do |format|
       if @book.save
-        format.html { redirect_to book_url(@book), notice: 'Book was successfully created.' }
+        format.html { redirect_to book_url(@book, locale: I18n.locale), notice: I18n.t('controllers.books.notice.create.success') }
```
url全てに`locale:`指定していってますが、一箇所で済むように出来ます。この辺読んでみましょう。
https://railsguides.jp/i18n.html#url-paramsを元にロケールを設定する

takao0206 replied:
アドバイス、ありがとうございます。
`application_controller.rb`のファイル等で、`default_url_options`を以下のようにオーバーライドすることで、ロケール情報もURLに渡せることが分かりました。勉強になりました。
```
 def default_url_options
  { locale: I18n.locale }
end
```

maedana commented on app/views/books/_book.html.erb:
```diff
@@ -1,21 +1,21 @@
 <div id="<%= dom_id book %>">
   <p>
-    <strong>Title:</strong>
+    <strong><%= t('activerecord.attributes.book.title') %>:</strong>
```
> [Rails 国際化 (i18n) API - Railsガイド](https://railsguides.jp/i18n.html) を読み、使えそうなメソッドがあれば積極的に使う

とあり、便利なものを利用したいですねー。より具体的には https://railsguides.jp/i18n.html#active-recordモデルで翻訳を行なう を読んでみましょう

takao0206 replied:
`Model.human_attribute_name(attribute)`メソッドのご紹介、ありがとうございました。
このメソッドを使った方が、コードが直感的になり、分かりやすいと思いました。

takao0206 committed:
ご指摘事項の修正（3点）

- `ja.yml`ファイルの中で不要な記述を削除した。
- ロケール情報をURLに追加するため、`default_url_options`をオーバーライドした。
- viewsでモデルの属性を翻訳をシンプルに表示できるように`Model.human_attribute_name(attribute)`メソッドを使用した。

takao0206 commented:
アドバイスいただき、ありがとうございました。
Rails i18nについて理解を深めることができました。
アドバイスをもとにコードを修正しました。
ご都合の良いときにご確認いただければ幸いです。
